### PR TITLE
Refactor PyTorch for explicit Lighnting/Wandb hyperparameters

### PR DESCRIPTION
**Pull request by mwalmsley**
Major changes
- Refactor pytorch define_model API to pass all hparams as explicit simple variables (e.g. architecture_name) rather than convenient but implicit functions (e.g. loss_function, model). This is messier in that it requires a lot of args, but allows for tracking/sweeps by wandb and restoring from checkpoints by lightning.
- Add stochastic_depth_probability as a hyperparameter. In short, this sometimes randomly skips whole blocks. Was previously set and on by default - should be no change to users, but can now be altered if desired.
- Add efficientnetb2 and b4 as options. No performance improvement on my current tests.
- Add predict_on_catalog.py for pytorch. Allows easy predictions on a catalog dataframe.

Minor (but potentially breaking) changes
- Rename pytorch model_architecture to architecture_name. Clearer that it is not the architecture func. itself. (may break api)
- Deprecate requirements.txt
- Set explicit lightning version (may break a build)
- Bump version ahead of pypi/pip

cc @patrikasvanagas, @camallen

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -26,7 +26,7 @@
             'torch == 1.10.1',
             'torchvision == 0.11.2',
             'torchaudio == 0.10.1',
-            'pytorch-lightning',
+            'pytorch-lightning==1.6.5',  # 1.7 requires protobuf version incompatible with tensorflow/tensorboard. Otherwise works.
```
Comment (camallen) [inline setup.py:29]: not blocking - in theory we could isolate the pytorch and tensorflow installs to their own python virtual envs, thus avoiding this conflict and allowing pytorch lightning to resolve as high as possible

Unless of course the tensorboard dependency i used in pytorch....then please ignore me.

File: zoobot/pytorch/training/train_with_pytorch_lightning.py [hunk 2]
```diff
@@ -184,36 +178,28 @@ def loss_func(preds, labels):  # pytorch convention is preds, labels
...
 
+    # TEMP
     trainer.fit(lightning_model, datamodule)
 
-    trainer.test(
-        model=lightning_model,
-        datamodule=datamodule,
-        ckpt_path='best'  # can optionally point to a specific checkpoint here e.g. "/share/nas2/walml/repos/gz-decals-classifiers/results/early_stopping_1xgpu_greyscale/checkpoints/epoch=26-step=16847.ckpt"
-    )
+    # trainer.test(
```
Comment (camallen) [inline zoobot/pytorch/training/train_with_pytorch_lightning.py:184]: any reason to remove the test step of after fitting? Does this speed up the system but impact quality?

**Review by camallen — APPROVED** [review]
LGTM - fwiw i've been using this version of the code in the batch processing system

File: setup.py (hunk 1 above)
Comment (mwalmsley) [reply setup.py:29]: > we could isolate the pytorch and tensorflow installs to their own python virtual envs

Good idea but I don't know how to do that. 

In practice there is very little advantage to being on the absolute latest package and I suspect the tensorboard team will update their deps shortly.

File: zoobot/pytorch/training/train_with_pytorch_lightning.py (hunk 2 above)
Comment (mwalmsley) [reply zoobot/pytorch/training/train_with_pytorch_lightning.py:185]: Testing routinely is bad practice as you may accidentally tune your hparams to overfit the model. I was being a bit lazy when adding here earlier. I have left it commented as example, with a warning note.
